applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -1014,11 +1009,26 @@ static void disable_subscription(struct subscriber *sub, uint8_t report_id)
 			continue;
 		}
 
+		if (sub->enabled_reports_bm == 0) {
+			/* Subscriber disconnected. */
+			send_empty_hid_out_reports(per, sub);
+		}
+
 		drop_enqueued_reports(&per->enqueued_reports, irep_idx);
 	}
 
 	/* And also this subscriber. */
 	drop_enqueued_reports(&sub->enqueued_reports, irep_idx);
+
+	if (sub->enabled_reports_bm == 0) {
+		/* Clear output reports list. */
+		for (size_t i = 0; i < ARRAY_SIZE(sub->out_reports); i++) {
+			if (sub->saved_out_reports_bm & BIT(i)) {
+				k_free(sub->out_reports[i].data);
```

Comment:
we shouldn't have need to use malloc. why can't data be stored in the array? We know how many output reports there are and report id is part of data.

Reply:
But we don't know the data size of a report until we receive such report. That's why `out_reports` (instance of `struct saved_out_report`) has only pointer to data and not the array itself. In order not to use malloc we would have to assume (or maybe w know that?) that data will not exceed certain size. For me it seems like an inefficient solution, because we would have to allocate maximum data size for all reports that can possibly appear.

Reply:
> But we don't know the data size of a report until we receive such report

That is not correct. We know what is the size of each output report. The same is true for input reports. We use dynamic data of event so we won't have to explicitly define event for each report.
At the moment we support only one output report - keyboard leds - and it takes if I am not mistaken 1 byte. (https://github.com/nrfconnect/sdk-nrf/blob/main/applications/nrf_desktop/configuration/common/hid_report_keyboard.h#L24)

This here could tell what reports are enabled https://github.com/nrfconnect/sdk-nrf/blob/main/applications/nrf_desktop/configuration/common/hid_report_desc.h#L58.